# Challenge 1 - Tic Tac Toe

In this lab you will perform deep learning analysis on a dataset of playing [Tic Tac Toe](https://en.wikipedia.org/wiki/Tic-tac-toe).

There are 9 grids in Tic Tac Toe that are coded as the following picture shows:

![Tic Tac Toe Grids](https://github.com/karimamzghi/lab-neural-networks/blob/master/your-code/tttboard.jpg?raw=1)

In the first 9 columns of the dataset you can find which marks (`x` or `o`) exist in the grids. If there is no mark in a certain grid, it is labeled as `b`. The last column is `class` which tells you whether Player X (who always moves first in Tic Tac Toe) wins in this configuration. Note that when `class` has the value `False`, it means either Player O wins the game or it ends up as a draw.

Follow the steps suggested below to conduct a neural network analysis using Tensorflow and Keras. You will build a deep learning model to predict whether Player X wins the game or not.

## Step 1: Data Engineering

This dataset is almost in the ready-to-use state so you do not need to worry about missing values and so on. Still, some simple data engineering is needed.

1. Read `tic-tac-toe.csv` into a dataframe.
1. Inspect the dataset. Determine if the dataset is reliable by eyeballing the data.
1. Convert the categorical values to numeric in all columns.
1. Separate the inputs and output.
1. Normalize the input data.

In [126]:
# importing the CSV file in file directory
import pandas as pd

df = pd.read_csv('/tic-tac-toe.csv')
df

,TL,TM,TR,ML,MM,MR,BL,BM,BR,class
0,x,x,x,x,o,o,x,o,o,True
1,x,x,x,x,o,o,o,x,o,True
2,x,x,x,x,o,o,o,o,x,True
3,x,x,x,x,o,o,o,b,b,True
4,x,x,x,x,o,o,b,o,b,True
...,...,...,...,...,...,...,...,...,...,...
953,o,x,x,x,o,o,o,x,x,False
954,o,x,o,x,x,o,x,o,x,False
955,o,x,o,x,o,x,x,o,x,False
956,o,x,o,o,x,x,x,o,x,False


In [127]:
# display first rows of the dataset
df.head()

,TL,TM,TR,ML,MM,MR,BL,BM,BR,class
0,x,x,x,x,o,o,x,o,o,True
1,x,x,x,x,o,o,o,x,o,True
2,x,x,x,x,o,o,o,o,x,True
3,x,x,x,x,o,o,o,b,b,True
4,x,x,x,x,o,o,b,o,b,True


In [128]:
# Check total features and rows
df.shape

(958, 10)

In [129]:
# Check dateType and non-null count and if there are missing data
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 958 entries, 0 to 957
Data columns (total 10 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   TL      958 non-null    object
 1   TM      958 non-null    object
 2   TR      958 non-null    object
 3   ML      958 non-null    object
 4   MM      958 non-null    object
 5   MR      958 non-null    object
 6   BL      958 non-null    object
 7   BM      958 non-null    object
 8   BR      958 non-null    object
 9   class   958 non-null    bool  
dtypes: bool(1), object(9)
memory usage: 68.4+ KB


We noticed there are 10 features non null features, 9 features are type objects and one type bool, no null objects


In [130]:
df.describe()

,TL,TM,TR,ML,MM,MR,BL,BM,BR,class
count,958,958,958,958,958,958,958,958,958,958
unique,3,3,3,3,3,3,3,3,3,2
top,x,x,x,x,x,x,x,x,x,True
freq,418,378,418,378,458,378,418,378,418,626


In [131]:
# Check missing values per column
df.isna().sum()

,0
TL,0
TM,0
TR,0
ML,0
MM,0
MR,0
BL,0
BM,0
BR,0
class,0


In [132]:
from numpy import astype
# Feature engineering
# We will apply to dummies encoding to the features TL through BR

df_encoded = pd.get_dummies(
    df,
    columns=['TL', 'TM', 'TR', 'ML', 'MM', 'MR', 'BL', 'BM', 'BR'],
    drop_first=True
).astype(int)

print(df_encoded)

     class  TL_o  TL_x  TM_o  TM_x  TR_o  TR_x  ML_o  ML_x  MM_o  MM_x  MR_o  \
0        1     0     1     0     1     0     1     0     1     1     0     1   
1        1     0     1     0     1     0     1     0     1     1     0     1   
2        1     0     1     0     1     0     1     0     1     1     0     1   
3        1     0     1     0     1     0     1     0     1     1     0     1   
4        1     0     1     0     1     0     1     0     1     1     0     1   
..     ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   
953      0     1     0     0     1     0     1     0     1     1     0     1   
954      0     1     0     0     1     1     0     0     1     0     1     1   
955      0     1     0     0     1     1     0     0     1     1     0     0   
956      0     1     0     0     1     1     0     1     0     0     1     0   
957      0     1     0     1     0     0     1     0     1     0     1     1   

     MR_x  BL_o  BL_x  BM_o  BM_x  BR_o

In [133]:
df_encoded["class"] = df_encoded["class"].astype(int)
df_encoded

,class,TL_o,TL_x,TM_o,TM_x,TR_o,TR_x,ML_o,ML_x,MM_o,MM_x,MR_o,MR_x,BL_o,BL_x,BM_o,BM_x,BR_o,BR_x
0,1,0,1,0,1,0,1,0,1,1,0,1,0,0,1,1,0,1,0
1,1,0,1,0,1,0,1,0,1,1,0,1,0,1,0,0,1,1,0
2,1,0,1,0,1,0,1,0,1,1,0,1,0,1,0,1,0,0,1
3,1,0,1,0,1,0,1,0,1,1,0,1,0,1,0,0,0,0,0
4,1,0,1,0,1,0,1,0,1,1,0,1,0,0,0,1,0,0,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
953,0,1,0,0,1,0,1,0,1,1,0,1,0,1,0,0,1,0,1
954,0,1,0,0,1,1,0,0,1,0,1,1,0,0,1,1,0,0,1
955,0,1,0,0,1,1,0,0,1,1,0,0,1,0,1,1,0,0,1
956,0,1,0,0,1,1,0,1,0,0,1,0,1,0,1,1,0,0,1


In [149]:
# Split Data into input and output
# X = features, y = target
X = df_encoded.drop(columns=["class"])
y = df_encoded["class"]

print("input: ", X.shape)
print("output: ", y.shape)

input:  (958, 18)
output:  (958,)


In [135]:
from sklearn.model_selection import train_test_split

# split our dataset into 80% train and 20% test

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

print("input train: ", X_train.shape)
print("output train: ", y_train.shape)
print("input test: ", X_test.shape)
print("output test: ", y_test.shape)

input train:  (766, 18)
output train:  (766,)
input test:  (192, 18)
output test:  (192,)


## Step 2: Build Neural Network

To build the neural network, you can refer to your own codes you wrote while following the [Deep Learning with Python, TensorFlow, and Keras tutorial](https://www.youtube.com/watch?v=wQ8BIBpya2k) in the lesson. It's pretty similar to what you will be doing in this lab.

1. Split the training and test data.
1. Create a `Sequential` model.
1. Add several layers to your model. Make sure you use ReLU as the activation function for the middle layers. Use Softmax for the output layer because each output has a single lable and all the label probabilities add up to 1.
1. Compile the model using `adam` as the optimizer and `sparse_categorical_crossentropy` as the loss function. For metrics, use `accuracy` for now.
1. Fit the training data.
1. Evaluate your neural network model with the test data.
1. Save your model as `tic-tac-toe.model`.

In [152]:
import tensorflow.keras
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Dropout
from tensorflow.keras.layers import Input
from tensorflow.keras.optimizers import Adam

num_classes = 1

seq_model = Sequential()
seq_model.add(Input(shape=(X_train.shape[1],)))
seq_model.add(Dense(64, activation="relu"))
seq_model.add(Dropout(0.2))

seq_model.add(Dense(32, activation='relu'))
seq_model.add(Dropout(0.2))

seq_model.add(Dense(16, activation="relu"))

# Define the output layer
seq_model.add(Dense(num_classes, activation='sigmoid'))

# Define the optimizer
my_opt = Adam(learning_rate=0.001)

seq_model.compile(
      loss='binary_crossentropy',
      metrics=['accuracy']
 )

In [153]:
# Display a summary of our deep Neural Network
seq_model.summary()

Model: "sequential_14"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense_48 (Dense)                │ (None, 64)             │         1,216 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_24 (Dropout)            │ (None, 64)             │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_49 (Dense)                │ (None, 32)             │         2,080 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_25 (Dropout)            │ (None, 32)             │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_50 (Dense)                │ (None, 16)             │           528 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_51 (Dense)                │ (None, 1)              │            17 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 3,841 (15.00 KB)

 Trainable params: 3,841 (15.00 KB)

 Non-trainable params: 0 (0.00 B)

In [154]:
# Train the model
batch_size = 32
epochs = 100

history = seq_model.fit(X_train, y_train,
                    batch_size=batch_size,
                    epochs=epochs,
                    verbose=1,
                    validation_data=(X_test, y_test))

Epoch 1/100
24/24 ━━━━━━━━━━━━━━━━━━━━ 1s 10ms/step - accuracy: 0.6123 - loss: 0.6639 - val_accuracy: 0.6510 - val_loss: 0.6107
Epoch 2/100
24/24 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.6554 - loss: 0.6129 - val_accuracy: 0.6562 - val_loss: 0.5870
Epoch 3/100
24/24 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.6789 - loss: 0.5886 - val_accuracy: 0.7083 - val_loss: 0.5597
Epoch 4/100
24/24 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7245 - loss: 0.5533 - val_accuracy: 0.7448 - val_loss: 0.5295
Epoch 5/100
24/24 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7585 - loss: 0.5273 - val_accuracy: 0.7656 - val_loss: 0.5023
Epoch 6/100
24/24 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7715 - loss: 0.4866 - val_accuracy: 0.7760 - val_loss: 0.4748
Epoch 7/100
24/24 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7820 - loss: 0.4805 - val_accuracy: 0.7969 - val_loss: 0.4427
Epoch 8/100
24/24 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8094 - loss: 0.4416 - val_accuracy: 0.7812 - 

In [159]:
# Evaluate the Model
score = seq_model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Test loss: 0.11572764068841934
Test accuracy: 0.9739583134651184


In [160]:
seq_model.save("tic-tac-toe.keras")

## Step 3: Make Predictions

Now load your saved model and use it to make predictions on a few random rows in the test dataset. Check if the predictions are correct.

In [164]:
from tensorflow.keras.models import load_model
import numpy as np

# Load the saved model
loaded_tic_tac_toe_model = load_model("tic-tac-toe.keras")

# Select 10 random rows from X_test
X_test_sample = X_test.sample(10, random_state=42)
y_test_sample = y_test.loc[X_test_sample.index]

# Make predictions
predictions = loaded_tic_tac_toe_model.predict(X_test_sample)

predicted_classes = np.argmax(predictions, axis=1)

results = X_test_sample.copy()

results["Actual: "] = y_test_sample.values
results["Predicted: "] = predicted_classes

print(results)

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 71ms/step
     TL_o  TL_x  TM_o  TM_x  TR_o  TR_x  ML_o  ML_x  MM_o  MM_x  MR_o  MR_x  \
943     0     1     0     1     1     0     1     0     1     0     0     1   
110     0     1     0     1     0     0     1     0     0     1     0     0   
88      0     1     0     1     1     0     1     0     0     1     1     0   
365     1     0     1     0     0     1     0     1     0     1     0     1   
718     0     1     0     0     1     0     0     1     0     1     1     0   
275     0     1     0     0     0     0     0     1     0     1     1     0   
542     0     0     1     0     0     1     1     0     0     1     0     0   
785     1     0     0     1     1     0     0     0     1     0     0     1   
30      0     1     0     1     0     1     1     0     1     0     0     0   
876     1     0     0     0     0     0     1     0     0     1     0     0   

     BL_o  BL_x  BM_o  BM_x  BR_o  BR_x  Actual:   Predicted:   
943     0     1     0     1

## Step 4: Improve Your Model

Did your model achieve low loss (<0.1) and high accuracy (>0.95)? If not, try to improve your model.

But how? There are so many things you can play with in Tensorflow and in the next challenge you'll learn about these things. But in this challenge, let's just do a few things to see if they will help.

* Add more layers to your model. If the data are complex you need more layers. But don't use more layers than you need. If adding more layers does not improve the model performance you don't need additional layers.
* Adjust the learning rate when you compile the model. This means you will create a custom `tf.keras.optimizers.Adam` instance where you specify the learning rate you want. Then pass the instance to `model.compile` as the optimizer.
    * `tf.keras.optimizers.Adam` [reference](https://www.tensorflow.org/api_docs/python/tf/keras/optimizers/Adam).
    * Don't worry if you don't understand what the learning rate does. You'll learn about it in the next challenge.
* Adjust the number of epochs when you fit the training data to the model. Your model performance continues to improve as you train more epochs. But eventually it will reach the ceiling and the performance will stay the same.

In [167]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Dropout, Input
from tensorflow.keras.optimizers import Adam

improved_model = Sequential()

improved_model.add(Input(shape=(X_train.shape[1],)))

improved_model.add(Dense(128, activation="relu"))
improved_model.add(Dropout(0.2))

improved_model.add(Dense(64, activation="relu"))
improved_model.add(Dropout(0.2))

improved_model.add(Dense(32, activation="relu"))
improved_model.add(Dropout(0.2))

improved_model.add(Dense(16, activation="relu"))

improved_model.add(Dense(num_classes, activation="sigmoid"))

my_opt = Adam(learning_rate=0.0005)

improved_model.compile(
    optimizer=my_opt,
    loss="binary_crossentropy",
    metrics=["accuracy"]
)

history = improved_model.fit(
    X_train,
    y_train,
    epochs,
    batch_size,
    verbose=1,
    validation_data=(X_test, y_test)
)

test_loss, test_accuracy = improved_model.evaluate(X_test, y_test)

print("Improved Test Loss: ", test_loss)
print("Improved Test Accuracy: ", test_accuracy)

Epoch 1/32
8/8 ━━━━━━━━━━━━━━━━━━━━ 3s 33ms/step - accuracy: 0.4530 - loss: 0.7080 - val_accuracy: 0.5625 - val_loss: 0.6847
Epoch 2/32
8/8 ━━━━━━━━━━━━━━━━━━━━ 0s 11ms/step - accuracy: 0.5862 - loss: 0.6780 - val_accuracy: 0.6562 - val_loss: 0.6600
Epoch 3/32
8/8 ━━━━━━━━━━━━━━━━━━━━ 0s 11ms/step - accuracy: 0.6462 - loss: 0.6572 - val_accuracy: 0.6562 - val_loss: 0.6432
Epoch 4/32
8/8 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step - accuracy: 0.6554 - loss: 0.6382 - val_accuracy: 0.6510 - val_loss: 0.6310
Epoch 5/32
8/8 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step - accuracy: 0.6606 - loss: 0.6262 - val_accuracy: 0.6615 - val_loss: 0.6190
Epoch 6/32
8/8 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step - accuracy: 0.6593 - loss: 0.6151 - val_accuracy: 0.6667 - val_loss: 0.6075
Epoch 7/32
8/8 ━━━━━━━━━━━━━━━━━━━━ 0s 11ms/step - accuracy: 0.6815 - loss: 0.6037 - val_accuracy: 0.6823 - val_loss: 0.5951
Epoch 8/32
8/8 ━━━━━━━━━━━━━━━━━━━━ 0s 11ms/step - accuracy: 0.6828 - loss: 0.5931 - val_accuracy: 0.7188 - val_loss: 0.5836


**Which approach(es) did you find helpful to improve your model performance?**

After increasing the model complexity by adding more hidden layers and neurons and adjusting the learning rate from 0,001 to 0,005, the improved model achieved a slightly higher loss (0.1349 vs. 0.1157) and slightly lower accuracy (96.88% vs. 97.40%) than the initial model. Therefore the original architecture was best fit for the Tic-Tac-Toe dataset. Adding more complexity did not improve the performance.